src/Middleware/RequestThrottling/src/RequestThrottlingMiddleware.cs

```diff
@@ -70,48 +49,43 @@ public RequestThrottlingMiddleware(RequestDelegate next, ILoggerFactory loggerFa
         /// <returns>A <see cref="Task"/> that completes when the request leaves.</returns>
         public async Task Invoke(HttpContext context)
         {
-            var waitInQueueTask = _requestQueue.TryEnterQueueAsync();
-
-            if (waitInQueueTask.IsCompletedSuccessfully && waitInQueueTask.Result)
-            {
-                RequestThrottlingLog.RequestRunImmediately(_logger, ActiveRequestCount);
-            }
-            else
-            {
-                RequestThrottlingLog.RequestEnqueued(_logger, ActiveRequestCount);
-                await waitInQueueTask;
-                RequestThrottlingLog.RequestDequeued(_logger, ActiveRequestCount);
-            }
-
-            if (!waitInQueueTask.Result)
-            {
-                RequestThrottlingLog.RequestRejectedQueueFull(_logger);
-                context.Response.StatusCode = StatusCodes.Status503ServiceUnavailable;
-                await _requestThrottlingOptions.OnRejected(context);
-                return;
-            }
+            Interlocked.Increment(ref _totalRequests);
 
             try
             {
-                await _next(context);
+                var success = await _queuePolicy.TryEnterAsync();
+                if (success)
+                {
+                    try
+                    {
+                        await _next(context);
+                    }
+                    finally
+                    {
+                        _queuePolicy.OnExit();
+                    }
+                }
+                else
+                {
+                    RequestThrottlingLog.RequestRejectedQueueFull(_logger);
+                    context.Response.StatusCode = StatusCodes.Status503ServiceUnavailable;
+                    await _onRejected(context);
+                }
             }
             finally
             {
-                _requestQueue.Release();
+                Interlocked.Decrement(ref _totalRequests);
             }
         }
-
+        
         /// <summary>
-        /// The number of requests currently on the server.
-        /// Cannot exceeed the sum of <see cref="RequestThrottlingOptions.RequestQueueLimit"> and </see>/><see cref="RequestThrottlingOptions.MaxConcurrentRequests"/>.
+        /// The total number of requests (queued and active) within the server
         /// </summary>
-        public int ActiveRequestCount
+        public int TotalRequestCount
```

Comment:
Hosting already know the TotalRequestCount. Counting the request currently sitting in the queue or the counting the requests currently executing is more interesting.

Reply:
Yeah I mostly choose `TotalRequests` because its convenient in tests. Switching to a `QueuedRequests` would be more useful, and then we could subtract from the Hosting value to get the active requests.